public/app/features/dashboard-scene/panel-edit/PanelEditor.tsx

```diff
@@ -81,8 +83,63 @@ export class PanelEditor extends SceneObjectBase<PanelEditorState> {
 
     const panelMngr = this.state.panelRef.resolve();
 
-    if (sourcePanel!.parent instanceof SceneGridItem) {
-      sourcePanel!.parent.setState({ body: panelMngr.state.panel.clone() });
+    const sourcePanelParent = sourcePanel!.parent;
+    const sourcePanelGrandparent = sourcePanelParent!.parent!;
+    if (!(sourcePanelGrandparent instanceof SceneGridLayout)) {
+      console.error('Expected grandparent to be SceneGridLayout!');
+      return;
+    }
+
+    const normalToRepeat = !sourcePanel!.state.repeat && panelMngr.state.panel.state.repeat;
+    const repeatToNormal = sourcePanel!.state.repeat && !panelMngr.state.panel.state.repeat;
+
+    if (sourcePanelParent instanceof SceneGridItem) {
+      if (normalToRepeat) {
+        const panel = panelMngr.state.panel;
+        const repeatDirection = panel.state.repeatDirection ?? 'h';
+        const repeater = new PanelRepeaterGridItem({
+          key: sourcePanelParent.state.key,
+          x: sourcePanelParent.state.x,
+          y: sourcePanelParent.state.y,
+          width: repeatDirection === 'h' ? 24 : sourcePanelParent.state.width,
+          height: sourcePanelParent.state.height,
+          itemHeight: sourcePanelParent.state.height,
+          source: panelMngr.state.panel.clone(),
+          variableName: panel.state.repeat!,
+          repeatedPanels: [],
+          repeatDirection: panel.state.repeatDirection,
+          maxPerRow: panel.state.maxPerRow,
+        });
+        sourcePanelGrandparent.setState({
+          children: sourcePanelGrandparent.state.children.map((child) =>
+            child.state.key === sourcePanelParent.state.key ? repeater : child
+          ),
+        });
+      } else {
+        sourcePanelParent.setState({ body: panelMngr.state.panel.clone() });
+      }
+    } else if (sourcePanelParent instanceof PanelRepeaterGridItem) {
+      if (repeatToNormal) {
+        const panelClone = panelMngr.state.panel.clone();
+        panelClone.setState({ repeat: undefined, repeatDirection: undefined, maxPerRow: undefined });
+        const gridItem = new SceneGridItem({
+          key: sourcePanelParent.state.key,
+          x: sourcePanelParent.state.x,
+          y: sourcePanelParent.state.y,
+          width: sourcePanel!.state.repeatDirection === 'h' ? 8 : sourcePanelParent.state.width,
+          height: sourcePanel!.state.repeatDirection === 'v' ? 8 : sourcePanelParent.state.height,
+          body: panelClone,
+        });
+        sourcePanelGrandparent.setState({
+          children: sourcePanelGrandparent.state.children.map((child) =>
+            child.state.key === sourcePanelParent.state.key ? gridItem : child
+          ),
+        });
+      } else {
+        sourcePanelParent.setState({ source: panelMngr.state.panel.clone() });
+      }
```

Comment:
I tested this part and it seems to be working just fine (editing and viewing).

The code looks clumsy, I agree. Faced the same problem with handling multiple variants of changing the data source, but we managed to fix that with structural simplification. I remember trying there an another abpproach by wrapping the objects in yet-another-abstraction to unify the handleing, but that was just delegating complexity to another place.

For now I would focus on trying to get the repeat options into the VizPanelManager as of Torkel's suggestion, and keeping this code as-is.